[GStreamer][MSE][CMake] Limit video support to 1080p@30fps on RaspberryPi

pgorszkowski-igalia opened the pull request:
#### 9d318e8396da78d13f59f006803d8d1a77a0d8b4
<pre>
[GStreamer][MSE][CMake] Limit video support to 1080p@30fps on RaspberryPi
<a href="https://bugs.webkit.org/show_bug.cgi?id=245270">https://bugs.webkit.org/show_bug.cgi?id=245270</a>

Reviewed by Adrian Perez de Castro and Xabier Rodriguez-Calvar.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::supportsType):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/254721@main">https://commits.webkit.org/254721@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19e42cc68e37f1d41d820612aa38ef9a3362648c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89972 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20618 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99300 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156817 "Found 2 new test failures: js/error-should-not-strong-reference-global-object.html, streams/readable-stream-create-after-worker-terminates-crash.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93978 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33011 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82307 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93612 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95620 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26226 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76772 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26143 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81094 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80906 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/69145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81600 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/34252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/14989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76488 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/32093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15930 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26480 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3316 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/35837 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38888 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79074 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/37737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35014 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17330 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -266,6 +266,7 @@ set(INTROSPECTION_INSTALL_TYPELIBDIR "${LIB_INSTALL_DIR}/girepository-1.0")
 SET_AND_EXPOSE_TO_BUILD(WTF_PLATFORM_QUARTZ ${ENABLE_QUARTZ_TARGET})
 SET_AND_EXPOSE_TO_BUILD(WTF_PLATFORM_X11 ${ENABLE_X11_TARGET})
 SET_AND_EXPOSE_TO_BUILD(WTF_PLATFORM_WAYLAND ${ENABLE_WAYLAND_TARGET})
+SET_AND_EXPOSE_TO_BUILD(WTF_PLATFORM_RPI ${USE_WPEWEBKIT_PLATFORM_RPI})
```
Where is `USE_WPEWEBKIT_PLATFORM_RPI` defined?

pgorszkowski-igalia replied:
Currently it is on the buildroot config level. I know that you want to have rather on cmake level - so recognize building target during configuration and set USE_WPEWEBKIT_PLATFORM_RPI if it is RPI.

aperezdc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -54,6 +54,22 @@
+// Raspberry Pi only supports up to 1080p@30fps hardware video decoding.
+#define MEDIA_MAX_WIDTH 1920.0f
+#define MEDIA_MAX_HEIGHT 1080.0f
+#define MEDIA_MAX_FRAMERATE 30.0f
```
There is nothing specific about the RaspberryPi here. The option can be useful for other platforms where the available hardware decoders (or the software based decoding) are not able to keep up with higher video resolutions.

I would prefer to call this option something like `USE(FULLHD_VIDEO_DECODING_LIMIT)`&mdash;if anybody else has a better name which transmits the same idea, that would be fine as well.

pgorszkowski-igalia replied:
I agree with you, it should be possible to use this settings also on other platforms (which support only limited resolutions), so as conclusion: I should change the name of the option to USE(FULLHD_VIDEO_DECODING_LIMIT) instead of PLATFORM(RPI).
The question is how it should be activated? Should it be set on upper level (buildroot, yocto config)?
Or in case RPi, detect it on cmake level and set for RPi (but what about other platform?)?

I think that detection of the supported playback resolution on the configuration level (cmake) can be not trivial task.

aperezdc replied:
What I meant with autodetection is to not have a CMake option, but make the C++ code detect the capabilities of the decoder and adapt at runtime without needing to rebuild.

At any rate, the CMake option must be specified explicitly by the build system or the person doing the build. We do *not* want to detect things at the CMake level: it would be very error prone, specially when cross-compilation is involved.

ThomasDevoogdt replied:
Buildroot can perfectly export this new general limit flag, and can be added to the the relevant device defconfigs. Would also put FHD as suffix, so that e.g. 4K can be added later.

aperezdc reviewed (changes requested):
I would very much prefer to detect at runtime whether the available hardware (or software) video decoders have some limits, instead of hardcoding maximums at build time. Also, I perfectly understand that doing it at runtime would be more complex, so I am willing to accept this as solution for now.

pgorszkowski-igalia committed:
[GStreamer][MSE][CMake] Limit video support to 1080p@30fps on RaspberryPi
https://bugs.webkit.org/show_bug.cgi?id=245270

Reviewed by Adrian Perez de Castro and Xabier Rodriguez-Calvar.

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::supportsType):
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/254721@main
